docs/concurrency_design.md

```diff
@@ -129,3 +129,90 @@ sequenceDiagram
 Some pending definitions:
 
 - The block tracker could eventually be a block cache, and maintain blocks and their state in an ETS that can be accessed easily by other processes.
+
+## Another possible design: per-block processes
+
+Conceptually, the core concurrency issue of the current design is that we use processes to represent steps instead of them representing independent pieces of computation/data. Most of the inter-process communication is call-based and sequential, effectively blocking processes while still adding the overhead of message passing and mailboxes with big objects like blocks and states.
+
+What we need is to have one process per independent task. In our case, our independent unit for is **the block processing**. We need to have a **single process per block**, that takes care of:
```

Comment:
so, you mean a single process per block, and not a separate process for block processing? That would only benefit us if we have a fork, in normal times (a straight line of blocks) it won't make a difference. But it makes sense to optimize for the worst scenario, not for the optimal one.

Reply:
Yes, this is one process per block, although there are several benefits:
1. It is true, works for the worst case scenario.
2. Naturally unblocks every step of the pipeline. That is, pending blocks won't block while it waits for a state transition to happen. The store or beacon chain won't block either.
3. If pending blocks is not blocked, then we can process all of the incoming gossip continuously, and discard the ones that are not valid early, without their validation depending on other blocks being processed. This also reduces the size of the processes mailbox.
4. Everything happens in parallel when needed (state transition, deserialization, decompression, etc).